site/docs/api/sources-sinks.md

```diff
@@ -894,7 +894,128 @@ The Kafka sink and source are based on version 2.2.0, this means Kafka
 connector will work with any client and broker having version equal to
 or greater than 1.0.0.
 
-### JMS
+### Amazon Kinesis
+
+[Amazon Kinesis Data
+Streams](https://aws.amazon.com/kinesis/data-streams/) (KDS) is a
+massively scalable and durable real-time data streaming service. All
+data items passing through it, called _records_, are assigned a
+_partition key_. As the name suggests, partition keys group related
+records together. Records with the same partition key are also ordered.
+Partition keys are grouped into _shards_, the base throughput unit of
+KDS. The input and output rates of shards is limited. Streams can be
+resharded at any time.
+
+To read from Kinesis, the only requirement is to provide a KDS stream
+name. (Kinesis does not handle deserialization itself, it only provides
+serialized binary data.)
+
+```java
+Pipeline p = Pipeline.create();
+p.readFrom(KinesisSources.kinesis(STREAM).build())
+  .withNativeTimestamps(0)
```

Comment:
Are we sure that timestamps within one shard are strictly time-ordered? Are they assigned by a single machine in AWS. If, inside AWS, some other machine takes over after a failed machine, could it have delayed clock? If this isn't explicitly specified by AWS, I wouldn't relay on 0 allowed lag.

Also, when a processor starts to consume a new shard, i guess it will happen with a delay after that shard was created. It might already contain data - but those will be rendered late.

Reply:
Since shards are meant to be the unit of parallelisation for Kinesis and since shards are "small" (meaning they have modest input/output data rates), I'm reasonably sure that all data ingestion of any one shard is handled by the same machine. I don't know how they handle failover, what happens if the shard migrates from one machine to another. But regardless, this is just an example, one can always specify a non-zero allowed lag.

Reply:
Of course one can fix our example, but an example should be exemplary. It's common that anomalies happen only when there are other anomalies, but that's exactly what guarantees are about - IMap is perfectly consistent, except in time of failures. Sitting presidents are kicked out of the internet only at times of a revolution.

Also you didn't address the issue with a new shard being rendered late. That will be tough one...